### DISCUSSION TITLE
Unsupported install code

### DISCUSSION THREAD

**Discussion by TautrimasM** [Koenkk/zigbee2mqtt#19982] [General]
Hi!

I am trying to add a device via install code because it does not have any pairing button. I have scanned the QR code and got the following line:

G$M:IC2%Z:540F57FFFE599FAA$I:**D79CB21C6D197CE7A3339A683A90DFF2442A**%M:1246 

I suppose that the bold part should be the install code, so I tried adding using it. I have also tried adding using the whole line. Both times I got this error:

"Request 'zigbee2mqtt/bridge/request/install_code/add' failed with error: 'Unsupported install code, got 72 chars, expected 95 or 91'"

I am fairly new to ZigBee, and as far as I was googling, the install code should be 32 characters. So I was wondering if anyone could enlighten me on what length it should be and how to connect my device to zigbee2mqtt. Maybe it should be converted somehow? Or maybe it is my zigbee adapter the problem? 

Additional info:
Device:  Danfoss Icon 2 Main Controller (supposed to support zigbee 3.0)
Zigbee Adapter: Zigbee CC2531 USB Dongle

**Comment by torablien** [comment]
Hey @TautrimasM, did you figure this out? I am running into a similar issue - scanning my different Zigbee devices' QR codes also gives varied lengths that aren't 95 or 91 (e.g. `Z:B0F7DEFFFE52C4DE$I:6F53FA0D037852A5EE8375E7607BD1A57E0B`, `Z:4CC206FFFE306C43$I:150A1A57D11A1A01362622DBA97ACF0F9185$D:202%B:4CC206306C43$P:983639%M:1220$F:002D`, `Z:4CC2061900002956$I:3BAF129043F34F3BBCAFF836982AB4BD820F$D:203%B:000000000000$P:000000%M:1220$F:0003`).  

It looks like the code that throws this is: https://github.com/Koenkk/zigbee-herdsman/blob/master/src/controller/controller.ts#L214-L231 which suggests it is parsing out the IEEE address and key. Based on the lengths, I believe that is the `Z:` and `I:` value respectively, so we could possibly work around this by just padding the remaining part of the install code with 0s so it accepts it and parses those two values. Alternatively, you can use the Aqara regex to effectively pass in the values: `G$M:-$S:-$D:-%Z$A:your-ieee-addr$I:your-install-code` since that pass bypasses the length checks.

@Koenkk does that makes sense in this context?

**Reply by TautrimasM** [reply]
Hello,

Yes, you are right, I just found out this a couple of days ago.
For future reference, I will note that I believe you found that info inside this discussion: [https://github.com/Koenkk/zigbee2mqtt/discussions/8498](https://github.com/Koenkk/zigbee2mqtt/discussions/8498)

Since it is discussed that there is no official standard for QR install codes, my quick fix for testing purposes was just to change RegEx  in the  installFolder/node_modules/zigbee-herdsman/dist/controller/controller.js
from:
installCode.match(/^G\$M:.+\$A:(.+)\$I:(.+)$/);
to:
installCode.match(/^G\$M:IC2%Z:(.+)\$I:(.+)%M:(\d+)$/);

P.S. I used ChatGPT to generate regex.

This is by no means a production-ready solution, but it was enough for me to stumble across another issue: The install code feature is not supported by my USB dongle as stated here: [https://github.com/Koenkk/zigbee2mqtt/discussions/19176](https://github.com/Koenkk/zigbee2mqtt/discussions/19176)

**Reply by torablien** [reply]
Thanks for the link to that discussion, I hadn't seen that before!

@Koenkk can we extend the implementation in `controller.ts` to accept different formats? Or an "advanced" option where we can just pass in the IEEE address and key directly instead given that manufacturers use varied formats?

**Reply by Koenkk** [reply]
@torablien yes, feel free to create a pull request to extend it.

**Reply by TautrimasM** [reply]
@torablien I suppose we could extend the addInstallCode function to try to parse input string as a json if parse as QR code failed, then we could pass installer code and ieee address bundled up as a json. What do you think?

**Reply by torablien** [reply]
yeah, if we don't want to edit the UI as well that should work

**Comment by Maxxis20** [comment]
Hi,

Did you get this to work? 

I got My Icon2 MC connected fine with just pushing the right button (Zigbee pairing).
(USB: HomeAssistant Sky Connect)
The data that is automatically populated is very limited.

Did you create a external definition for the MC already?

**Reply by TautrimasM** [reply]
Hi, 

Yes, I did connect the way you have described. 

And yes, i have made an external converter for Icon2 MC and RT, and it successfully gets the following parameters:

- MC link quality
- Thermostat battery
- Humidity
- Temperature
- Setpoint
- Setpoint limits
- Setpoint change source 
- ChildLock

however, i was not able to get the heating status of the zones, it seems that clusters are not supported by MC, i asked Danfoss if this thing is available and am currently waiting for a reply. 

I think i will add this converter to the project when i get an answer from Danfoss. And for now, feel free to play with it:
[Icon2MC.zip](https://github.com/Koenkk/zigbee2mqtt/files/13986591/Icon2MC.zip)

**Reply by Maxxis20** [reply]
Thank you, looks similar to what i did by modifying the original danfoss.ts converter.
I will give this a try and inplement more points. 

Check out this also for detailed Cluster and point information on the Icon2-MC:https://filetransfer.io/data-package/6LZoVuHY#link

**Reply by Maxxis20** [reply]
Did you connect the RT's to the MC? And used the Icon2 App for setup and assigning RT to room?
I tried your .js file and ran into the same problem as previously, the room sensors report with the same MC device number (0x0210), and has a endpoint and group number that identifies the thermostat. 
How would one separate the thermostats for the MQTT message? Because now if you have several thermostats the previous values are overwritten.

**Comment by Maxxis20** [comment]
> I am trying to add a device via install code because it does not have any pairing button. I have scanned the QR code and got the following line:
> Device: Danfoss Icon 2 Main Controller 

On the Icon2 controller the right (OK) button works as a Zigbee pairing button.
If the left button is pressed so the light goes from Blinking-Steady-Blinking, the device is removed from network, so no factory reset needed. This needs to be done for endpoints to be updated (if sensors have been added after the ZB pairing).

**Comment by TautrimasM** [comment]
Yes, I did.

MC endpoints correspond to thermostats. If you want to read specific
thermostat then you should send request to specific endpoint.

Also, you can read the complete state, then parameters of each thermostat
will have suffix of the endpoint that the thermostat is located, i.e.
battery_1, occupied_setpoint_2.

P. S. Device number you mention is model identifier, each individual MC has
the same, it has nothing to do with thermostats.

**Reply by Maxxis20** [reply]
So i added the converter to the configuration.yaml:
"  external_converters:
    - 0x0210.js"
and created a corresponding file to "/zigbee2mqtt/0x0210.js" which is you code exactly.

and then im stuck on getting the data out in a good form.
The debug gives the endpoint and group but the mqtt publish (info) gives just the values out.

So how do you sort the thermostat values to individual values, because the mqtt topic is only the MC IEEE Address..

Z2M gives this output:
Debug: 2024-01-28 13:01:00Received Zigbee message from '0xf4b3b1fffedd851a', type 'attributeReport', cluster 'msRelativeHumidity', data '{"measuredValue":8300}' from endpoint 3 with groupID 0

Info 2024-01-28 13:01:00MQTT publish: topic 'zigbee2mqtt/0xf4b3b1fffedd851a', payload '{"battery":77,"humidity":83,"linkquality":224,"temperature":-327.68}'

**Comment by TautrimasM** [comment]
It seems that your setup is not picking up the external converter file that
i have provided.

Did you put it into ./data folder?

Also, provide logs if there are any error messages, it will be easier to
diagnose.

**Reply by Maxxis20** [reply]
Thanks for your reply!
Apparently when running Z2M inside HAOS you need to also add the external converter in settings.
(Z2M > Settings > External converters > "+", and give the file name including ".js" and submit and restart.
This is missed in the Z2M instructions, but could be a HAOS bug..
Could also be because i run unsupported/beta HAOS on Pi5..

Now i get the thermostat numbers, even though the second one is not _2 but _l2 for some reason..
But this i can work with in Node-Red.

    "local_temperature_1": 23.5,
    "local_temperature_3": 24.91,
    "local_temperature_l2": 24.82,
    "max_heat_setpoint_limit_1": 35,
    "max_heat_setpoint_limit_3": 35,
    "max_heat_setpoint_limit_l2": 35,

**Reply by Maxxis20** [reply]
And when the MC is added it configures correctly but after 10min it drops back to "Support status Not supported" for some reason.. 

Yours says "Supported" right? @TautrimasM 

But now the state info works.
And sometimes when i configure the MC after a factory reset it identifies as a thermostat which is very weird..

![Screenshot 2024-01-29 at 22 44 24](https://github.com/Koenkk/zigbee2mqtt/assets/111085242/189db87b-5b5f-41b4-bd55-edefbe80b877)

And now using it like this.. 
![Screenshot 2024-01-29 at 23 34 33](https://github.com/Koenkk/zigbee2mqtt/assets/111085242/3417e5e1-3a1c-4b44-8c45-6ebd2f6a21fa)

**Comment by TautrimasM** [comment]
I noticed that usually this happens if you try to read model id of any of
the thermostat endpoints. If this happen, then read model id of the
endpoint 232 (this represents mc itself) and it will fix itself.

**Reply by Maxxis20** [reply]
Thanks, i made a modification here to dont read the genBasic-modelId:
"// await endpoint.read('genBasic', ['modelId', 'powerSource']);
   await endpoint.read('genBasic', ['powerSource']);"
And now the MC has stayed Supported since.

**Comment by TautrimasM** [comment]
Great notion, thanks.

**Reply by Maxxis20** [reply]
Hi,

Have you succeeded with writing a setpoint to a thermostat?

I have tried everything from different tutorials/directories/examples..
The problem seems to be that the Set value is not finding its correct spot because enpoints have own names for the variables.

I get an error:
Received MQTT message on 'zigbee2mqtt/0-----------a/set' with data '{"occupied_heating_setpoint_1":15.5}'
Error No converter available for 'occupied_heating_setpoint_1' (15.5)

**Reply by Maxxis20** [reply]
My latest .js file: 
[DfMF_Latest_BU.zip](https://github.com/Koenkk/zigbee2mqtt/files/14142738/DfMF_Latest_BU.zip)

**Reply by TautrimasM** [reply]
I managed to set setpoint and childlock (locks the thermostat buttons) using these payload and topic:

Setpoint:
payload = { "occupied_heating_setpoint": 22}

childlock:
payload = { "keypad_lockout": "lock1"}
payload = { "keypad_lockout": "unlock"}

topic = 'zigbee2mqtt/friendly_name/end_Point/set'

**Reply by Maxxis20** [reply]
I also got it to work yesterday, by analyzing the errors and checking other tutorials.
Thanks!
Now my next "problem" is to get the floor sensor to show, it is in the 'msTemperatureMeasurement' cluster  under 'measuredValue'.
I can read it in the dev console, but cant seem to get it to add to the converter, it wont show up under the state menu in Z2M..

**Reply by mmirannalti** [reply]
Hi both.
I really need your expert advice.
I have in my new house a Danfoss Icon2 MC (088U2100) and a thermostat (088U2121).
I have installed them following the instructions, connecting the thermostat to the MC using the setup app provided by Danfoss.
Everything works as expected.

Now I would like to integrate the Danfoss Icon2 environment into my Smart Home system (Z2M + HA).
The official Z2M documentation is not very clear but following your posts, I understand that it should be possible to pair the MC with Z2M, just using the right button to place the MC in pairing mode.
I have tried but although the led is starting blinking, the pairing process is not initiated by Z2M.
Is there anything I am missing? Do I need to provide an installation code? Do I need some custom converter?

Thanks for your help!

**Reply by Maxxis20** [reply]
@mmirannalti did you click the "Permit Join" in Z2M GUI?

**Reply by mmirannalti** [reply]
Yes, I did.

**Reply by Maxxis20** [reply]
And you dont have it connected to the Ally Gateway?
Pressing the right button "OK" should make the left light blink in light blue.
And it will pop up in Z2M if you have Permit Join on.

**Reply by mmirannalti** [reply]
It worked! I was pressing+holding. Pressing was enough :-)
Thanks!!!

I see the set temperature, the measured temperature and the zone status. 
It seems I can also change the value for the set temperature.
Actually all I needed!

The umidity percentage I can see on the thermostat is not exposed.

**Comment by akauppi** [comment]
I tried to type in (manually, not QR) an install code from a Schneider/Wiser smoke alarm, today.

The codes themselves, without URL coatings, are 8+1 16-bit groups, the +1 being a checksum (that for all purposes of a user is part of the "install code"). This made 36 chars of hex, white space removed. [Silicon Labs Application Note 1089](https://www.silabs.com/documents/public/application-notes/an1089-using-installation-codes-with-zigbee-devices.pdf) is helpful, here.

![image](https://github.com/Koenkk/zigbee2mqtt/assets/1297734/d8c9f14d-fe51-4001-8334-1be254449dab)

My work-around?

Publish this. Subscribe. Push the button for pairing, this time. :)

**Reply by Maxxis20** [reply]
Hmm. There is no pairing button on the Wieser stuff?

**Reply by akauppi** [reply]
Sure there is. https://www.productinfo.schneider-electric.com/wiser_eu/wiser-smoke-alarm-230-v_device-user-guide_wiser_se/English/Wiser%20Smoke%20Alarm%20230%20V_WSE_Device%20user%20guide_0001066064.xml/$/PairingDeviceManually_WSE_SmokeAlarm-BatteryTSK_0000854799

It’s marked ”S/R” and at the bottom of the device. I got things to pair. Just wanted to try with the installation codes, first.

**Comment by JohnConnett** [comment]
I am investigating new device support for the [Philips Hue Secure Floodlight Camera EU](https://github.com/Koenkk/zigbee2mqtt/issues/21650).

The base has a sticker with a six hex digit number and a **QR Code** of the following shape: `HUE:Z:0123456789ABCDEF0123456789ABCDEF0123 M:0123456789ABCDEF D:L3B A:1184`. Not the actual values for the 'Z:' and 'M:" hex strings! The actual 'Z:' is consistent with a 128-bit value and 16-bit CRC format given in **Base Device Behavior Specification Version 1.0**, Zigbee Document 13-0402-13 and **PRO Base Device Behavior Specification, v3.0.1**, Zigbee Document 16-02828-012. The CRCs match both for the example in the specifications and for the actual 'Z:'. The 'M:' value is the same as the IEEE Address for the floodlight. Strong possibility that these are the values required for an install code commissioning.

The Hue Secure wired camera Ownership card has a six hex digit Setup code; a 12 hex digit Device ID; and a **QR Code** with the shape `MT:AB012CD.34EFGH5IJ6K7-L8M9NO.P0QRS12TUVW34X5Y6Z7-A8-BC9012D3456EF78-GH9012I345JK67LMN8OPQR9S-012T-UV3WXYZ4A5BC6DEF.7GH-I8J901`. Again, not the actual values. Decoded as [Base36](https://en.wikipedia.org/wiki/Base36) none of the component parts look likely to be a 144-bit install code. Other decoding suggestions gratefully received! As the camera part is capable of Bluetooth LE, WiFi and Zigbee some of the contents are likely to be for the other supported wireless types.

My hope is that I can configure the Bluetooth LE and WiFi using the Hue App but still be able to find a mechanism for joining it to my existing Zigbee network.

Is there any further information on using install codes with Zigbee2MQTT? I have looked but not found much information.

**Comment by JohnConnett** [comment]
> Hey @TautrimasM, did you figure this out? I am running into a similar issue - scanning my different Zigbee devices' QR codes also gives varied lengths that aren't 95 or 91 (e.g. `Z:B0F7DEFFFE52C4DE$I:6F53FA0D037852A5EE8375E7607BD1A57E0B`, `Z:4CC206FFFE306C43$I:150A1A57D11A1A01362622DBA97ACF0F9185$D:202%B:4CC206306C43$P:983639%M:1220$F:002D`, `Z:4CC2061900002956$I:3BAF129043F34F3BBCAFF836982AB4BD820F$D:203%B:000000000000$P:000000%M:1220$F:0003`).

I think that the CRCs on all three of your install codes may be incorrect. Assuming that the non-CRC parts are correct I think that:

1. `6F53FA0D037852A5EE8375E7607BD1A5-7E0B` should be `6F53FA0D037852A5EE8375E7607BD1A5-FA5C`
2. `50A1A57D11A1A01362622DBA97ACF0F-9185` should be `50A1A57D11A1A01362622DBA97ACF0F-0CCA`
3. `3BAF129043F34F3BBCAFF836982AB4BD-820F` should be `3BAF129043F34F3BBCAFF836982AB4BD-60F3`

The install code in the original post from @TautrimasM, `D79CB21C6D197CE7A3339A683A90DFF2442A` looks good.

Thought I would mention it in case your install codes are rejected. Were all three devices from the same manufacturer?

**Reply by JohnConnett** [reply]
Here's the the [Rust](https://www.rust-lang.org/) program I wrote to check the CRCs:
```
use crc::{Crc, Algorithm};

fn main() {
    // use custom algorithm
    const CUSTOM_ALG: Algorithm<u16> = Algorithm {
        width: 16,
        poly: 0x1021,
        init: 0xffff,
        refin: true,
        refout: true,
        xorout: 0xffff,
        check: 0x906e,
        residue: 0x0000
    };

    let crc = Crc::<u16>::new(&CUSTOM_ALG);

    // 3132 3334 3536 3738 39
    assert_eq!(crc.checksum(b"123456789"), crc.algorithm.check);
    let v: Vec<u8> = vec![0x31, 0x32, 0x33, 0x34,
                          0x35, 0x36, 0x37, 0x38,
                          0x39];
    assert_eq!(crc.checksum(&v), crc.algorithm.check);

    // Example install code from Zigbee Specification
    // 83FE D340 7A93 9723 A5C6 39B2 6916 D505 C3B5
    let v: Vec<u8> = vec![0x83, 0xfe, 0xd3, 0x40,
                          0x7a, 0x93, 0x97, 0x23,
                          0xa5, 0xc6, 0x39, 0xb2,
                          0x69, 0x16, 0xd5, 0x05];
    assert_eq!(crc.checksum(&v), 0xb5c3);

    // Example from https://github.com/Koenkk/zigbee2mqtt/discussions/19982#discussioncomment-8178764
    // D79C B21C 6D19 7CE7 A333 9A68 3A90 DFF2 442A
    let v: Vec<u8> = vec![0xd7, 0x9c, 0xb2, 0x1c,
                          0x6d, 0x19, 0x7c, 0xe7,
                          0xa3, 0x33, 0x9a, 0x68,
                          0x3a, 0x90, 0xdf, 0xf2];
    assert_eq!(crc.checksum(&v), 0x2a44);

    // Example from https://github.com/Koenkk/zigbee2mqtt/discussions/19982#discussioncomment-8178764
    // 6F53 FA0D 0378 52A5 EE83 75E7 607B D1A5 *7E0B*
    let v: Vec<u8> = vec![0x6f, 0x53, 0xfa, 0x0d,
                          0x03, 0x78, 0x52, 0xa5,
                          0xee, 0x83, 0x75, 0xe7,
                          0x60, 0x7b, 0xd1, 0xa5];
    assert_eq!(crc.checksum(&v), 0x5cfa);

    // Example from https://github.com/Koenkk/zigbee2mqtt/discussions/19982#discussioncomment-8178764
    // 150A 1A57 D11A 1A01 3626 22DB A97A CF0F *9185*
    let v: Vec<u8> = vec![0x15, 0x0a, 0x1a, 0x57,
                          0xd1, 0x1a, 0x1a, 0x01,
                          0x36, 0x26, 0x22, 0xdb,
                          0xa9, 0x7a, 0xcf, 0x0f];
    assert_eq!(crc.checksum(&v), 0xca0c);

    // Example from https://github.com/Koenkk/zigbee2mqtt/discussions/19982#discussioncomment-8178764
    // 3BAF 1290 43F3 4F3B BCAF F836 982A B4BD *820F*
    let v: Vec<u8> = vec![0x3b, 0xaf, 0x12, 0x90,
                          0x43, 0xf3, 0x4f, 0x3b,
                          0xbc, 0xaf, 0xf8, 0x36,
                          0x98, 0x2a, 0xb4, 0xbd];
    assert_eq!(crc.checksum(&v), 0xf360);
}

```

**Comment by JohnConnett** [comment]
Based on a sample of five strings decoded from **QR Codes** on Hue devices (floodlight, three different models of bulb, dimmer switch) they can be matched using this regular expression: `/^HUE:Z:([0-9A-F]{36}) M:([0-9A-F]{16}) D:[0-9A-Z]+ A:\d{4}$/`.

The decoded strings are composed of:
 
    1. 'HUE:'                           // Hue header
    2. 'Z:'                             // Z tag
    3. 36 uppercase hexadecimal digits  // 128-bit install code + 16-bit CRC
    4. ' '                              // single space
    5. 'M:'                             // M tag
    6. 16 uppercase hexadecimal digits  // IEEE Address
    7. ' '                              // single space
    8. 'D:'                             // D tag
    9. digits or uppercase letters      // 3 or 5 of them?
    10. ' '                             // single space
    11. 'A:'                            // A tag
    12. 4 digits                        // always 4 of them?

The meanings of the 'D:' and 'A:' values are unknown. The regular expression is strict so that exceptions will be identified to allow the regular expression to be further refined. It rejects some obvious validation errors such as invalid characters and invalid lengths.

Hopefully, other formats used for install codes will be sufficiently distinctive that a more complex parser will be unnecessary! The  Danfoss Icon 2 Main Controller example given above has a 'G$M:' header. The (as yet undeciphered) Hue Secure camera has 127 characters with a 'MT:' header. If other formats have an easily identified header then they can have their own individual regular expressions or parsers. It would also allow for more precise error messages when an install code is rejected.

I also have a Hue wall switch module which does not have a **QR Code**. The instructions in [How to set up a Hue wall switch module Connect with a Hue Bridge](https://www.philips-hue.com/en-us/support/connect-hue-product/accessories/wall-switch-module) do not mention a **QR Code**.

Is anyone collating the different styles of decoded **QR Codes** for install codes?

**Reply by JohnConnett** [reply]
I have successfully commissioned a Hue bulb using **ZHA** with a **SkyConnect** coordinator. This was achieved by going to [http://homeassistant.local:8123/developer-tools/service](http://homeassistant.local:8123/developer-tools/service), selecting the service **Zigbee Home Automation: Permit**, filling in the **Source IEEE** and **Install code** fields then clicking **CALL SERVICE**.

I also tried filling in the **QR code** but that was rejected with _Failed to call service zha.permit. couldn't convert qr code: [...]_.

Is something similar available in Zigbee2MQTT, perhaps via the WEB UI?

**Reply by JohnConnett** [reply]
The **ZHA** code for handling **QR Codes** is: [https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/core/helpers.py#L340-L395](https://github.com/home-assistant/core/blob/dev/homeassistant/components/zha/core/helpers.py#L340-L395).

It has regular expressions for Consciot, Enbrighten, Aqara and Bosch formats. Neither the Danfoss nor Hue formats would be matched.

**Comment by JohnConnett** [comment]
An install code add can be triggered using MQTT with the topic `zigbee2mqtt/bridge/request/install_code/add`.

One way of achieving this is using the MQTT integration. Navigate to **Settings -> Devices & Services -> MQTT -> Configure**. Under **Publish a packet** enter the topic `zigbee2mqtt/bridge/install_code/add` and enter the install code as the payload.

I used a (fake) Aqura install code of `G$M: $A:0123456789ABCDEF$I:83FED3407A939723A5C639B26916D505C3B5`. The following line appeared in the log:

```
info  2024-03-20 10:46:50: MQTT publish: topic 'zigbee2mqtt/bridge/response/install_code/add', payload '{"data":{"value":"G$M: $A:0123456789ABCDEF$I:83FED3407A939723A5C639B26916D505C3B5"},"status":"ok"}'
```
Same with a (fake) 95 character install code of `xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0123456789ABCDEFxxx83FED3407A939723A5C639B26916D505C3B5` when the following appeared in the log:

```
info  2024-03-20 11:47:40: MQTT publish: topic 'zigbee2mqtt/bridge/response/install_code/add', payload '{"data":{"value":"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0123456789ABCDEFxxx83FED3407A939723A5C639B26916D505C3B5"},"status":"ok"}'
```

So if I can find a mechanism to obtain an install code for my Hue Secure camera I will have something to try ...

**Comment by JohnConnett** [comment]
Has anyone successfully joined a device using an Install Code string?

I removed a **Hue white ambiance E27 1600lm with Bluetooth** (actually B22 not E27) and tried to join it using Aqara and 95 character Install Code strings created using the IEEE address and install code for the bulb.

Two attempts with each. All four of them received `"status":"ok"` responses. Bulb didn't rejoin.

Selected `Permit join(All)` and it rejoined straight away.

Log attached of just lines containing `001788010bc22490`, the IEEE address of the bulb.

[light.txt](https://github.com/Koenkk/zigbee2mqtt/files/14758409/light.txt)

**Comment by JohnConnett** [comment]
Just added another toy to my collection, the new [ubisys Switching Actuator S1-R (Series 2)](https://www.ubisys.de/en/products/light-and-consumers/switching-actuator-s1-r/) which I hope to use as a smart switch for my immersion heater. Might be a while before it is installed as that will require a suitably qualified electrician.

This has another style of unsupported install code. The text equivalent of the QR Code is: **Z:0102030405060708$I:0102030405060708090A0B0C0D0E0F1090FD%G$M:S1-R%M:10F2**.

Note this is not the actual code from my device but the EUI-64: **0102030405060708** and I/C: **0102030405060708090A0B0C0D0E0F1090FD** have the right form including the I/C checksum: **90FD**.

The [ubisys S1-R-2](https://www.zigbee2mqtt.io/devices/S1-R-2.html#ubisys-s1-r-2) is in the Supported Devices list.

This is the second **ubisys** device to support **Zigbee Direct**. I asked them if they had any tools that supported **Zigbee Direct**. Their reply contained:  _regarding tools, I suggest using our free of charge ubisys Smart Home App which is a ZVD – albeit only for iOS for the time being_. However, I don't have any iOS devices ...

The factory reset procedure is:

_1. The device has to be powered up for at least 10 seconds._
_2. Power off for 10 seconds._
_3. Power on for 10 seconds._

_Repeat steps 2.-3. in the aforementioned **10-second rhythm** another 3 times, and with the last sequence keep the device powered up._

I can see that **Zigbee Direct** commissioning could be really useful!

Still interested to know if anyone successfully joined any device using an Install Code string?

**Comment by JohnConnett** [comment]
Zigbee2MQTT version: 2.1.3-1
[Electrolama zzh Multiprotocol RF Stick (CC2652R1 - External Antenna)](https://electrolama.com/products/zzh-multiprotocol-rf-stick) with [CC2652R_coordinator_20240710.hex](https://github.com/Koenkk/Z-Stack-firmware/releases/tag/Z-Stack_3.x.0_coordinator_20240710) firmware.

Successfully joined the same **Hue white ambiance E27 1600lm with Bluetooth** (actually B22 not E27) using the (fake) 95 character install code that was unsuccessful before!

Not sure what has changed to make it work. However, I have successfully joined a device using an Install Code string.

**Reply by JohnConnett** [reply]
I may have been too hasty claiming success at least in the case of Z-Stack adapters. [zigbee-herdsman/src/adapter/z-stack/adapter/zStackAdapter.ts](https://github.com/Koenkk/zigbee-herdsman/blob/master/src/adapter/z-stack/adapter/zStackAdapter.ts) contains the following code starting at line 769:
```
    public async addInstallCode(ieeeAddress: string, key: Buffer, hashed: boolean): Promise<void> {
        assert(this.version.product !== ZnpVersion.zStack12, 'Install code is not supported for ZStack 1.2 adapter');
        // TODO: always use 0x2? => const hashedKey = hashed ? key : ZSpec.Utils.aes128MmoHash(key);
        const payload = {installCodeFormat: hashed ? 0x2 : 0x1, ieeeaddr: ieeeAddress, installCode: key};
        await this.znp.request(Subsystem.APP_CNF, 'bdbAddInstallCode', payload);
    }
```
Note that in the commented out line starting '// TODO:' the install code is _not_ hashed using the Matyas-Meyer-Oseas (MMO) hash function. I _think_ this will result in a different value for the 128-bit pre-configured link-key (which should be derived from the 144-bit install code using the MMO hash function).

This _may_ cause problems with a backup using a Z-Stack adapter followed by a restore using an adapter that expects the MMO hash function to have been used. That problem could also occur when the second adapter is a Z-Stack adapter after a fix has been applied to use the MMO hash function.

I also note that in the case of the ember adapter the MMO hash function is also used _to prevent a compromise of the backup data from compromising the current link keys_ (This is not used for the ezsp or dconz adapters). While this is possibly a sensible security feature it may not be compatible with the [Open ZigBee Coordinator Backup Format](https://github.com/zigpy/open-coordinator-backup).

There appear to be some inconsistencies in the handling of add device using a 144-bit install code!

**Reply by JohnConnett** [reply]
Raised as an issue: [Inconsistencies in add device with install code #1353](https://github.com/Koenkk/zigbee-herdsman/issues/1353).
